src/robotlibcore.py

```diff
@@ -141,6 +131,29 @@ def get_keyword_documentation(self, name):
             doc = '{}\n\n{}'.format(doc, tags) if doc else tags
         return doc
 
+    def get_keyword_types(self, name):
+        kw = self.keywords[name]
+        if hasattr(kw, 'robot_types'):
+            robot_types = getattr(kw, 'robot_types')
+            if robot_types:
+                return robot_types
+        default_types = self._get_default_types(kw)
+        if default_types:
+            return default_types
+        return {} if PY2 else kw.__annotations__
+
+    def _get_default_types(self, kw):
+        args, defaults, varargs, kwargs = self._get_arg_spec(kw)
+        default_types = {}
+        for name, value in defaults:
+            default_types[name] = type(value)
+        if varargs:
+            default_types[varargs] = list
+        if kwargs:
+            default_types[kwargs] = dict
+        if default_types:
+            return default_types
```

Comment:
I'm not sure should DynamicCore look at defaults at all. Robot considers type information it gets via defaults to be implicit and info it gets via `@keyword`, annotations or `get_keyword_types` to be explicit. Depending on the type, conversion can differ between implicit and explicit type information. Generally implicit conversion is less strict.

Somewhat unfortunately the dynamic API only supports explicit types. This means that if DynamicCore returns some type information it's always considered explicit, even if RF itself would have considered it implicit. That means Robot can do conversion that shouldn't be done. I believe it's better for DynamicCore to be scripts and add this functionality later if needed.

Possible exceptions to the above are Boolean values and `None`. The reason I'd be ready to handle them differently than other default values are:
- Their implicit and explicit conversion doesn't differ. See [typeconverters.py](https://github.com/robotframework/robotframework/blob/master/src/robot/running/arguments/typeconverters.py) for details.
- They are probably the most commonly used default values.

Even with Booleans and `None` I'd consider implementing this support separately after getting the basic functionality ready.

Reply:
Can you add code for annotations in _get_default_types().
_get_arg_spec() also should return 'annotations' for this change.
```python
def _get_default_types(self, kw):
        args, annotations, defaults, varargs, kwargs = self._get_arg_spec(kw)
        default_types = {}
        for arg in args:
            if arg in annotations:
                default_types[arg] = annotations[arg]
        for name, value in defaults:
            if name in annotations:
                default_types[name] = annotations[name]
            else:
                default_types[name] = type(value)
        if varargs:
            default_types[varargs] = list
        if kwargs:
            default_types[kwargs] = dict
        if default_types:
            return default_types
```